File: rikiki/bin/t.js
//
// This software is dedicated to the public domain.
//

function call(f, ...args)
{
	return f(...args);
}

function include(file)
{
	switch(file == "std") {
	case true: {
	}};
}

function struct(v,cls)
{
}

function array(n)
{
	return new Array(n);
}

function bytes(n)
{
	return new Uint8Array(n);
}

function _(txt)
{
	return (new TextEncoder()).encode(txt);
}

function poke(b,i,v)
{
	var n = v;
	if (typeof v === "string") {
		n = v.charCodeAt(0);
	}
	b[i] = n;
}

function peek(b,i)
{
	return b[i];
}

function print(v)
{
	var s = (new TextDecoder()).decode(v);
	console.log(s);
}
function print2(v)
{
	console.log(parseInt(v));
}

function free(o)
{
	delete o;
}

function file_size(f)
{
	return 0;
}

function file_load(f,s)
{
}

//
// This software is dedicated to the public domain.
//


// There is no type checking. Every variable is threated equaly as an 
// integer, a reference, a one dimension array of integer/reference, a 
// reference to a byte array or a function reference.

include("std"); // include an other source file

const MY_CONST = 47; //define a constant

class myst { // a structure with four members
	one; 
	next;
	three;
	four;
}; 

function test(){}; //Functions must be declared before calling them
function t2(a,b){}; //everything after // is a comment till the end of line

function startup(argc, argv) 
{		            // definition of the entry point function

	var i,a,s,b,end; //declaration of local variables
	var n;
	var f,y;
	struct(s,myst); // declare type of variable s to class "myst".

	a = array(123);   //allocate an array of 123 intergers
	s = new myst();   //allocate a "myst" structure
	b = bytes(3);     //allocate 3 bytes of memory
	
	poke(b,0,'B');	  // put a B in the first byte of b memory
	poke(b,1,'\n');
	poke(b,2,0);
	print(_("B:"));
	print(b);

	f = t2;		 // function pointer
	call(f, 100,2,3,4,5); // indirect function call
	f = f();

	n = 1 * 20 % 2 / 3 + (17 - 4); //that's all the operators that we have
	print2(n);

	s.one = 10001; // to access a struct member we 
                            //put the name of the struct 
                            //then the member name

	s.next = _("two"); // string literal (nul terminated as in C)
	s.three = 3;    // integer constant
	s.four = '4';   // character constant (translated to integer)

	i = a[1];	     // array access

	a[0] = _("yo");
	a[122] = _("lAst");

	print(_("héllo world!\n")); // UTF-8
	print2(peek(a[122], 1));  // print the integer value of the second byte
                                  // of the string pointed by the 123 cell of a
	print(_("\n"));
	print(argv[argc-1]);
	print2(argv[argc-1]);

	i = 0;
	y = i;
	while (i == y && i != y && i < y && i <= y && i > y && i >= y) {
	};

	switch (argc > 2) {      // switch is used as a replacement of if/else
	case true:	
		print(_("yo \n"));
		i = 2;
		end = 0;
		while (i < argc && (end == 0 || 3 == 2)) { // only while loop
			print(_("."));
			i = i + 1;
			switch (i) {
			case +10: 
				end = 1;
				break;
			};
		};
		break;
	case false:
	case -1:
		print(s.next);
		break;
	};
	free(a);
	test(); 	// function call
	free(s);
	f = file_size(argv[1]);
	switch(f > 0) { // a compare operation result is either +1, -1 or 0. 
	case -5: 
	case 2: 
	case 3:       // these are garbage, but are allowed
	case true: 		// +1 is true
		f = file_load(argv[1], f);
		print(f);
		break;
	default: 
	case false: break;      // -1 or 0 are false
	};
	return 0;
}

// real definition of functions
function test()
{
	print(_("test\n")); //print a string
	print2(MY_CONST); // print a number (print3 is for ternary encoded numbers)
	t2(1,2);
}

function t2( a , b ) {
	print2(a+b);
	print(_("\n"));
}

//
// This software is dedicated to the public domain.
//

switch (typeof process !== 'undefined') {
case true:
	startup(2,[_("theApp"), _("ARGG1"), 0]);break;
}

File: rikiki/bin/t.c
/*

   This software is dedicated to the public domain.

*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#ifndef var
#define var long
#endif

#define peek(buf, at) ((var)(((unsigned char*)buf)[at]))
#define poke(buf, at, v) ((unsigned char*)buf)[at] = (v)

var startup(var a, var b);

int main(int argc, char *argv[])
{
	return startup(argc, (var)(void*)argv);
}

var print(var str)
{
	printf("%s", (char*)str);
	return 0;
}

var print2(var n)
{
	printf("%ld", n);
	return 0;
}

var file_size(var path)
{
	FILE *fp;
	var si;
	fp = fopen((char*)path, "rb");
	if (!fp) {
		return 0;
	}
	fseek(fp, 0, SEEK_END);
	si = ftell(fp);
	fclose(fp);
	return si;
}

var file_load(var path, var size)
{
	char *buf;
	FILE *fp;
	var ret;

	fp = fopen((char*)path, "rb");
	if (!fp) {
		return 0;
	}
	buf = malloc(size+1);
	if (!buf) {
		return 0;
	}
	ret = fread(buf, 1, size, fp);
	if (ret != size) {
		free(buf);
		buf = 0;
	}
	buf[size] = '\0';
	fclose(fp);
	return (var)buf;
}

var file_save(var path, var size, var buf)
{
	FILE *fp;
	var ret;

	fp = fopen((char*)path, "wb");
	if (!fp) {
		return -1;
	}
	ret = fwrite((void*)buf, 1, size, fp);
	fclose(fp);
	return ret;
}


#define strcmp(a,b) strcmp((void*)a, (void*)b)
#define strdup(a) strdup((void*)a)

var print(var string);
var print2(var binary_number);
var print3(var ternary_number);
var ter2bin(var bin);
var bin2ter(var ter);
var and2(var a, var b);
var or2(var a, var b);
var xor2(var a, var b);
var inv2(var a);
var cons3(var a, var b);
var any3(var a, var b);
var xor3(var a, var b);
var neg3(var a);
#define MY_CONST 47
struct myst {
	var one;
	var next;
	var three;
	var four;
};
var test(void);
var t2(var a, var b);
var startup(var argc, var argv)
{
	var i;
	var a;
	var s;
	var b;
	var end;
	var n;
	var f;
	var y;
	a = (var)(void*)(malloc((123) * sizeof(var)));
	s = (var)(void*)(malloc(sizeof(struct myst)));
	b = (var)(void*)(malloc(((3) & ~(sizeof(var) - 1)) + sizeof(var)));
	poke(b, 0, 'B');
	poke(b, 1, '\n');
	poke(b, 2, 0);
	print(b);
	f = (var)(void*)(t2);
	((var (*)())f)(100, 2, 3, 4, 5);
	f = (var)(void*)(((var (*)())f)());
	n = (var)(void*)(1 * 20 % 2 / 3 + (17 - 4));
	print2(n);
	(((struct myst*)s)->one) = (var)(void*)(10001);
	(((struct myst*)s)->next) = (var)(void*)(((var)(void*)"two"));
	(((struct myst*)s)->three) = (var)(void*)(3);
	(((struct myst*)s)->four) = (var)(void*)('4');
	i = (var)(void*)((((var*)a)[1]));
	(((var*)a)[0]) = (var)(void*)(((var)(void*)"yo"));
	(((var*)a)[122]) = (var)(void*)(((var)(void*)"lAst"));
	print(((var)(void*)"hello world\n"));
	print2(peek((((var*)a)[122]), 1));
	print(((var)(void*)"\n"));
	print((((var*)argv)[argc - 1]));
	print2((((var*)argv)[argc - 1]));
	i = (var)(void*)(0);
	y = (var)(void*)(i);
	while (i == y && i != y && i < y && i <= y && i > y && i >= y) {
	};
	switch ((var)(argc > 2)) {
	case +1: {
		print(((var)(void*)"yo \n"));
		i = (var)(void*)(2);
		end = (var)(void*)(0);
		while (i < argc && (end == 0 || 3 == 2)) {
			print(((var)(void*)"."));
			i = (var)(void*)(i + 1);
			switch ((var)(i)) {
			case +10: {
				end = (var)(void*)(1);
				break;
			}
			};
		};
		break;
	}
	case 0:
	case -1: {
		print((((struct myst*)s)->next));
		break;
	}
	};
	free((void*)a);
	test();
	free((void*)(((struct myst*)s)));
	f = (var)(void*)(file_size((((var*)argv)[1])));
	switch ((var)(f > 0)) {
	case -5:
	case 2:
	case 3:
	case +1: {
		f = (var)(void*)(file_load((((var*)argv)[1]), f));
		print(f);
		break;
	}
	case -1:
	case 0: {
		break;
	}
	};
	return 0;
}
var test(void)
{
	print(((var)(void*)"test\n"));
	print2(47);
	t2(1, 2);
	return 0;
}
var t2(var a, var b)
{
	print2(a + b);
	print(((var)(void*)"\n"));
	return 0;
}
